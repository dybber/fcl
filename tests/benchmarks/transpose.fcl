sig transpose : forall 'a. int -> int -> ['a] -> ['a]
fun transpose cols rows elems =
  generate (cols * rows)
           (fn n =>
              let i = n / rows in
              let j = n % rows
              in index elems (j * rows + i))

-- Split a 2D array (represented as a flat vector) into a 2D array of 2D arrays
-- First argument @splitSize is width and height of tiles
-- Second and third argument is the width and height of the original array
sig splitGrid : forall 'a. int -> int -> int -> int -> int -> ['a] -> [['a]]
fun splitGrid splitSize tileSize width groupsWidth groupsHeight elems =
  let mkTile =
        (fn groupIDy =>
          fn groupIDx =>
             generate tileSize
                      (fn tid =>
                         let localIDx = tid % splitSize in -- column in inner array
                         let localIDy = tid / splitSize in -- row in inner array
                         let xIndex = groupIDy * splitSize + localIDx in
                         let yIndex = groupIDx * splitSize + localIDy in
                         let ix = yIndex * width + xIndex
                         in index elems ix))
  in generate (groupsWidth * groupsHeight)
       (fn gid =>
          let groupIDy = gid / groupsWidth in -- row in outer
          let groupIDx = gid % groupsWidth -- column in outer
          in mkTile groupIDy groupIDx)

-- Concatenate a 2D array of 2D arrays
-- First argument @splitSize is the width and height of inner arrays
-- Second argument @tileSize must be === splitSize*splitSize
-- Third argument @groupsPerRow is the width of the outer array
-- Total number of elements is thus splitSize*splitSize*groupsPerRow*groupsPerRow
sig concatGrid : forall <lvl> 'a. int -> int -> int -> int -> [Program <lvl> ['a]<lvl>] -> Program <1+lvl> ['a]<1+lvl>
fun concatGrid <lvl>
               splitSize  -- width&height of inner array (2)
               tileSize
               groupsWidth       -- width of outer array (2)
               height
               arr =
    let mkGrid =
          (fn sh =>
             let gid = fst sh in -- index into outer (0..3)
             let tid = snd sh in -- index into inner (0..3)
             let groupIDx = gid % groupsWidth in -- column in outer
             let groupIDy = gid / groupsWidth in -- row in outer
             let localIDx = tid % splitSize in -- column in inner array
             let localIDy = tid / splitSize in -- row in inner array
             let xIndex = groupIDy * splitSize + localIDx in
             let yIndex = groupIDx * splitSize + localIDy in
             let ix = yIndex * height + xIndex
             in ix)
    in interleave <lvl> tileSize mkGrid arr

-- Transpose:
--  - split up in tiles
--  - transpose each tile
--  - force each tile to shared memory
--  - transpose outer array
--  - write back to global memory with concatGrid
sig transposeChunked : forall 'a. int -> int -> int -> ['a] -> Program <grid> ['a]<grid>
fun transposeChunked splitSize width height elems =
  let tileSize = splitSize * splitSize in
  let groupsWidth = width / splitSize in -- width of outer array
  let groupsHeight = height / splitSize in -- height of outer array
    splitGrid splitSize tileSize width groupsWidth groupsHeight elems
     |> mapPull (fn arr =>
                   do <block>
                    { b <- force <block> (push <block> arr)
                    ; return <block> (transpose splitSize splitSize b)
                    })
     |> transpose groupsWidth groupsHeight
     |> mapPull (fn arr => do <block>
                            { b <- arr
                            ; return <block> (push <block> b)
                            })
     |> concatGrid <block> splitSize tileSize groupsWidth height

sig transposeChunked16 : forall 'a. int -> int -> ['a] -> Program <grid> ['a]<grid>
val transposeChunked16 = transposeChunked 16

--- Testing
sig make_NxN_Matrix : int -> [int]
fun make_NxN_Matrix n = generate (n / n) (fn i => i)


sig example : Program <grid> [int]<grid>
val example = transposeChunked 16 4096 4096 (make_NxN_Matrix 4096)

sig main : Program <grid> unit
val main =
  do <grid>
   { arr <- example
   ; forceAndPrint 100 arr
   ; benchmark 100 (bind <grid> example (force <grid>))
   }
