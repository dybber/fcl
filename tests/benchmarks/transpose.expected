[0,4096,8192,12288,16384,20480,24576,28672,32768,36864,40960,45056,49152,53248,57344,61440,65536,69632,73728,77824,81920,86016,90112,94208,98304,102400,106496,110592,114688,118784,122880,126976,131072,135168,139264,143360,147456,151552,155648,159744,163840,167936,172032,176128,180224,184320,188416,192512,196608,200704,204800,208896,212992,217088,221184,225280,229376,233472,237568,241664,245760,249856,253952,258048,262144,266240,270336,274432,278528,282624,286720,290816,294912,299008,303104,307200,311296,315392,319488,323584,327680,331776,335872,339968,344064,348160,352256,356352,360448,364544,368640,372736,376832,380928,385024,389120,393216,397312,401408,405504]
